### ISSUE TITLE
go-restful returns invalid HTTP/1.0 responses for some errors

### ISSUE THREAD

**Issue by brackendawson** [emicklei/go-restful#438]
I found another spec thing. If you send a HTTP/1.0 request to go-restful it can return some invalid HTTP/1.0 responses, for example when the method is not allowed:

```
$ curl -XPOST -D- localhost:8080/hello -0
HTTP/1.0 405 Method Not Allowed
Allow: GET
Date: Sun, 21 Jun 2020 20:28:01 GMT
Content-Length: 23
Content-Type: text/plain; charset=utf-8

405: Method Not Allowed
```

405 is not listed as an error in RFC 1945: https://tools.ietf.org/html/rfc1945#section-9.4

I think the following HTTP/1.1 errors may be sent to HTTP/1.0 clients by the code as it stands today:
* 405 Method Not Allowed
* 406 Not Acceptable
* 410 Gone
* 415 Unsupported Media Type

Possibly I'm being too much of a stickler and nothing need be done about this, especially given the low likelihood of a HTTP/1.0 client relying on this error behaviour actually existing today.

We could put a check in the return path to downgrade any 4xx error code over 404 to a 400 when the response is a HTTP/1.0 response. The body can still hold the text of the HTTP/1.1 error because RFC 1945 9.2 does say "the server should include an entity containing an explanation of the error situation".

Some other web frameworks always return a 404 instead of 405 and 406 but RFC 1945 defines a 404 as the request-URI not being found, RFC 7231 expands this definition to include representations.

Another approach I've seen is to always respond to a HTTP/1.0 request with a `HTTP/1.1 426 Upgrade Required` response but that's not great either. The right way to tell a HTTP/1.0 client that would be in the body of a 400 error.

What do you think?

**Comment by emicklei** [comment]
hi,

thank you for sharing this finding and explaining what happens.
to be honest, I never thought about supporting HTTP/1.0 and was also not aware for the implications.
I am happy to include a change to fix the problem for your case; I prefer to follow the spec where possible.

**Comment by brackendawson** [comment]
I've been putting off doing a PR for this because I can't convince myself it's the right thing to do.

There's a couple of good reasons I see not to make a change here:
- It's additional code and additional behaviour, which we would have to document and maintain. It's an additional burden to users of the framework to understand, that HTTP/1.0 clients might see different errors generated by go-restful.
- While RFC1945 doesn't have errors between 404 and 500, other published drafts for HTTP/1.0 do, [this](https://www.w3.org/Protocols/HTTP/1.0/draft-ietf-http-v10-spec-01.html) one goes up to 411. I don't feel comfortable making assumptions about what version of version 1.0 the client is implementing.

I think this issue should be closed.

**Comment by emicklei** [comment]
thank you for sharing your final thoughs
